src/core/qgstiledownloadmanager.cpp

```diff
@@ -196,7 +239,14 @@ QgsTileDownloadManagerReply *QgsTileDownloadManager::get( const QNetworkRequest
 
     QObject::connect( entry.objWorker, &QgsTileDownloadManagerReplyWorkerObject::finished, reply, &QgsTileDownloadManagerReply::requestFinished );  // should be queued connection
 
-    addEntry( entry );
+    if ( isCachedRangeRequest( request ) )
+    {
+      QTimer::singleShot( 0, entry.objWorker, &QgsTileDownloadManagerReplyWorkerObject::replyFinished );
+    }
+    else
+    {
+      addEntry( entry );
+    }
   }
```

Comment:
Hmm I find the current changes quite hard to understand, they make the code more complex because of the change of existing invariants:
- it is expected that each new entry (with its worker object) gets added to the queue - this is not true anymore
- it is expected that every QgsTileDownloadManagerReplyWorkerObject has a valid reply object - also not true anymore

Changing existing invariants is generally not bad, but here it seems that it just increases complexity and there's another way to do it...

Here's my suggestion how to make the code flow simpler:
- the general idea is that for cached range requests, we wouldn't create new queue entries (and their worker objects) at all, because they are not really needed
- in QgsTileDownloadManager::get(), once we create QgsTileDownloadManagerReply object, let's check whether it is a cached range request (before a call to findEntryForRequest()) - if it is cached, let's do QTimer::singleShot() call, but it will go to a new private slot in QgsTileDownloadManagerReply - e.g. cachedRangeRequestFinished() - which would simply just read the cached entry and emit finished() signal.
- the code to add/remove entries in the queue would stay unchanged and in QgsTileDownloadManagerReplyWorkerObject::replyFinished() we would only keep the bit of code that stores successful requests with 206 http status.

Reply:
Great suggestions, implemented in the latest commit